### DISCUSSION TITLE
Way to customize the BentOs location

### DISCUSSION THREAD

**Discussion by callmedada** [bentoml/BentoML#2409] [Q&A]
Hey, BentoML team!

Happy Monday. I am currently trying to automate the workflow to customize the dockerfile to suit our strict deployment policy, and customized the openapi.ymal before containerize the BentOS. I am doing it by editing the Dockerfile and ymal files in the BentOS folders. I am wondering if we can make the BentOS in a customized folder but not under the current user folder (by default)? And if there is a better approach to customized the Dockerfile and OpenAPI, please let me know.  Any comments or criticize will be much appreciated. 

Many thanks in advance,
Dennis

**Comment by yubozhao** [comment]
Hey @callmedada 

This is an interesting question.

From what I understand:
You are trying to customize the generated Dockerfile and the openapi.yaml file in the bento directory.  And you are wondering could you customize where these files are generated. 
Did I get it right?

Right now, I don't think right now there is a good way to customize where these files are generated.  

I think the best way right now's write script for post bento building command for your customization.  The paths to these files are static. Let me know if I can help with that.

I would love to learn and understand more about the strict deployment policy.  We would love to help you to make BentoML more compliance with it.

Also feel free to message me on [slack](https://l.linklyhq.com/l/ktOh) or email: bo@bentoml.com, if there are sensitive info.

**Comment by parano** [comment]
Hi @callmedada - In general, we don't encourage users to change files under a Bento after it is build. 

As suggested here https://github.com/bentoml/BentoML/discussions/2367#discussioncomment-2547474, customizing the generated OpenAPI docs is currently not possible - you can only customize at the IO descriptor level, e.g. adding sample input or adding schema to a JSON input, but not the entire OpenAPI document.

BentoML does offer a number of ways to customize the docker base image. for example, you can choose a distro among "debian", "amazonlinux2", "alpine", "ubi8", and "ubi7", you can choose to include GPU related libraries or not, and you can set a `setup_script` to run upon docker image build. You can also use the `base_image` option to let BentoML use your own custom docker base image. All of those, doesn't not require users to change things under a Bento folder.  May I ask what kind of customization on the docker side that you need?

If customizing files under a Bento directory is needed and you want to customize where the Bento folder is generated, you can use the `export/import` API to export a Bento to your choice of path, after the Bento is built. See `bentoml export --help`.

**Comment by callmedada** [comment]
Hey @yubozhao  and @parano ,

Thanks a million for you guys answers and thoughts. Yes, basically we are trying to use our own artifactory to replace pip, and perform   multi-stage build (build from multiple staged images) to hide some sensitive information from the layers . Also, the default build seems have permission issue if not run as root in openshift. (Please see this issue : https://github.com/bentoml/BentoML/issues/1032, it still exists). Grant permission is not allowed by us, so I have to customize the dockerfile in order to run it in redshift. 

Thanks,
Dennis
